libkbfs/tlf_handle.go

```diff
@@ -931,40 +919,78 @@ func ParseTlfHandle(
 	if len(extensionSuffix) != 0 {
 		extensions, err = ParseTlfHandleExtensionSuffix(extensionSuffix)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
 	h, err := makeTlfHandleHelper(ctx, public, writers, readers, extensions)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	if !public {
 		currentUsername, currentUID, err := kbpki.GetCurrentUserInfo(ctx)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 
 		if !h.IsReader(currentUID) {
-			return nil, ReadAccessError{currentUsername, h.GetCanonicalName(), public}
+			return nil, nil, ReadAccessError{currentUsername, h.GetCanonicalName(), public}
 		}
 	}
 
-	if string(h.GetCanonicalName()) == name {
-		// Name is already canonical (i.e., all usernames and
-		// no assertions) so we can delay the identify until
-		// the node is actually used.
-		return h, nil
-	}
+	switch identifyBehavior {
+	case keybase1.TLFIdentifyBehavior_DEFAULT_KBFS:
+		if string(h.GetCanonicalName()) == name {
+			// Name is already canonical (i.e., all usernames and
+			// no assertions) so we can delay the identify until
+			// the node is actually used.
+			return h, nil, nil
+		}
 
-	// Otherwise, identify before returning the canonical name.
-	err = identifyHandle(ctx, kbpki, kbpki, h)
-	if err != nil {
-		return nil, err
+		// Otherwise, identify before returning the canonical name.
+		err = identifyHandle(ctx, kbpki, kbpki, h)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return nil, nil, TlfNameNotCanonical{name, string(h.GetCanonicalName())}
+	case keybase1.TLFIdentifyBehavior_CHAT:
+		breaks, err := identifyHandleWithIdentifyBehavior(ctx, kbpki, kbpki, h,
+			keybase1.TLFIdentifyBehavior_CHAT)
+		if err != nil {
+			return nil, breaks, err
+		}
+
+		if string(h.GetCanonicalName()) != name {
```

Comment:
> Ah, I didn't realize that; somehow I was thinking that was the only place that runs identify. Although what's the purpose of running identify here originally (before returning canonical name if it's not canonical)?

The issue as I recall (@akalin-keybase should correct me if I'm wrong) is that the resolve calls (which turn an `@` assertion into a username) don't actually check the proofs, it only relies on the Keybase database.  So if there were any assertions in the original name, we have to do an identify here to make sure it is safe to return the canonicalized name.

> 1) what should the behavior be like if the TLF is first time being accessed? Should identify fail to avoid creating it at all, or go ahead and create it but just don't set identifyDone to true, so that next time a normal KBFS operation happens, it'd run identify with KBFS behavior again?

Not sure I follow your question.  If `IdentifyForChat` fails, then the folder should definitely not be created.  If `IdentifyForChat` succeeds but with track breaks, then KBFS still needs to do a regular identify later -- whether or not you make a `folderBranchOps` or not doesn't make much difference to me.  If `IdentifyForChat` succeeds without any track breaks, then you can create a `folderBranchOps` and set `identifyDone=true`.

> 2) if the answer for 1) is latter, is it what "And when creating a new folder, you'd have to plumb the behavior through to folderBranchOps.SetInitialHeadToNew. " is about? If the answer for 1) is former, could you explain that a bit?

`GetTLFCryptKeys` can end up calling `folderBranchOps.SetInitialHeadToNew` if the TLF doesn't exist yet on the servers, and that will eventually call `Identify2`.  If you don't plumb anything through, it will do the regular identify, which you don't want.  If the folder already existed on the servers, you should add an explicit identify in `getOrInitializeNewMDMaster` when it is being called for chat purposes.